dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala

```diff
@@ -29,83 +29,172 @@ class Recurrent[T : ClassTag] (
+      case _ => throw new IllegalArgumentException("Cell not implemented")
+    }
+  }
+
+  /**
+   * Clone N models; N depends on the time dimension of the input
+   * @param size
+   */
+  private def extend(size: Int): Unit = {
+    require(modules != null && modules.length >= 1,
+      "Recurrent extend: should contain at least one module")
+    var t = modules.length
+    var flag = false
+    if (t < size) {
+      flag = true
+    }
+    while (t < size) {
+      super.add(modules(0).cloneModule())
+      t += 1
+    }
+    if (flag) {
+       share()
+    }
+  }
+
+  /**
+   * Sharing weights, gradWeights across all the modules in time dim
+   */
+  def share(): Unit = {
+    val params = modules(0).parameters()
+    var i = 1
+    while (i < modules.length) {
+      val curParams = modules(i).parameters()
+      var j = 0
+      while (j < curParams._1.length) {
+        curParams._1(j).set(params._1(j))
+        j += 1
+      }
+      j = 0
+      while (j < curParams._2.length) {
+        curParams._2(j).set(params._2(j))
+        j += 1
+      }
+      i += 1
+    }
+  }
+
+  private def hidInit(size: Array[Int]): Unit = {
+    if (hidden.isInstanceOf[Tensor[T]]) {
+      hidden.asInstanceOf[Tensor[T]].resize(size)
+    } else {
+      var j = 1
+      while (j <= hidden.asInstanceOf[Table].length) {
+        hidden.asInstanceOf[Table](j).asInstanceOf[Tensor[T]].resize(size)
+        j += 1
+      }
+    }
+  }
+
+  /**
+   * cloning inputs for reusing in backward
+   * @param tbl
+   * @return
+   */
+  private def shallowCopy(tbl: Table): Table = {
+    val ntbl = T()
+    var i = 1
+    while (i <= tbl.length) {
+      ntbl(i) = Tensor[T]().resizeAs(tbl(i)).copy(tbl(i))
+      i += 1
+    }
+    ntbl
+  }
+
   override def updateOutput(input: Tensor[T]): Tensor[T] = {
     require(input.dim == 3,
       "Recurrent: input should be a 3D Tensor, e.g [batch, times, nDim], " +
         s"current input.dim = ${input.dim}")
-    require(modules.length == 2,
-      "Recurrent: rnn container must include a cell and a non-linear layer, " +
-        s"current container length is ${modules.length}")
-
-    module = modules(0)
-    transform = modules(1)
 
-    batchSize = input.size(1)
-    times = input.size(2)
+    batchSize = input.size(batchDim)
+    times = input.size(timeDim)
+    extend(times)
 
     output.resize(Array(batchSize, times, hiddenSize))
-    hidden.resize(Array(batchSize, times + 1, hiddenSize))
 
+    hidInit(Array(batchSize, hiddenSize))
+    var currentOutput = T(hidden)
     var i = 1
     while (i <= times) {
-      val curInput = T(input.select(2, i), hidden.select(2, i))
-      val currentOutput = module.updateOutput(curInput)
-      transform.updateOutput(currentOutput)
-      output.select(2, i).copy(transform.output.toTensor)
-      hidden.select(2, i + 1).copy(transform.output.toTensor)
+      currentOutput(inputDim) = input.select(timeDim, i)
+      inputs(i) = shallowCopy(currentOutput)
```

Comment:
Why do you need inputs? It's just T(modules(i).output(inputDim), input.select(timeDim, i))